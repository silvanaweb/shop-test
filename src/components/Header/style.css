.Header {
  padding: var(--l-size) 0;
}

.Header__content {
  display: flex;
  justify-content: space-between;
  flex-direction: column-reverse;
}

.Header__link {
  color: var(--dark-grey);
  font-size: var(--m-size);
  font-weight: 500;
  text-decoration: none;
  text-transform: uppercase;
  margin-right: 1rem;
}
.Header__link:hover {
  color: var(--teal);
}
.Header__link.active {
  color: var(--teal);
}

@media (min-width: 45rem) {
  .Header {
    padding: var(--xl-size) 0;
  }
  .Header__content {
    flex-direction: row;
    align-items: center;
  }
  .Header__link {
    font-size: var(--font-medium);
  }
}
