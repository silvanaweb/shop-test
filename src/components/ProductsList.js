import React from "react";
import { connect } from "react-redux";
import selectProducts from "../store/selectors/products";

const ProductsList = props => {
  console.log(props.products);
  return (
    <div className="content-container">
      <div>List</div>
    </div>
  );
};

export { ProductsList };

const mapStateToProps = state => {
  console.log("state", state);
  console.log("ProductsList", state.products);
  return {
    products: selectProducts(state.products)
  };
};

export default connect(mapStateToProps)(ProductsList);
