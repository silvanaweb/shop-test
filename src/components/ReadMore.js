import PropTypes from "prop-types";
import React, { Component } from "react";
import Truncate from "react-truncate";

class ReadMore extends Component {
  constructor(...args) {
    super(...args);

    this.state = {
      expanded: false,
      truncated: false
    };

    this.handleTruncate = this.handleTruncate.bind(this);
    this.toggleLines = this.toggleLines.bind(this);
  }

  handleTruncate(truncated) {
    if (this.state.truncated !== truncated) {
      this.setState({
        truncated
      });
    }
  }

  toggleLines(event) {
    event.preventDefault();

    this.setState({
      expanded: !this.state.expanded
    });
  }

  render() {
    const { children, more, less, lines } = this.props;

    const { expanded, truncated } = this.state;

    return (
      <div>
        <Truncate
          lines={!expanded && lines}
          ellipsis={
            <span>
              ...{" "}
              <button className="Button--link" onClick={this.toggleLines}>
                {more}
              </button>
            </span>
          }
          onTruncate={this.handleTruncate}
        >
          {children}
        </Truncate>
        {!truncated && expanded && (
          <span>
            {" "}
            <button className="Button--link" onClick={this.toggleLines}>
              {less}
            </button>
          </span>
        )}
      </div>
    );
  }
}

ReadMore.defaultProps = {
  lines: 3,
  more: "read more",
  less: "show less"
};

ReadMore.propTypes = {
  children: PropTypes.node.isRequired,
  lines: PropTypes.number,
  less: PropTypes.string,
  more: PropTypes.string
};

export default ReadMore;
