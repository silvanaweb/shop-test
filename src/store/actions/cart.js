import { decreaseAmount } from "./products";

const addProductToCart = cartProduct => ({
  type: "ADD_PRODUCT_TO_CART",
  cartProduct
});

const startAddProductToCart = (product, qty) => {
  console.log("to cart", product, qty);
  return dispatch => {
    dispatch(decreaseAmount(product, qty));
    const cartProduct = {
      title: product.title,
      qty,
      price: product.price,
      image: product.image
    };
    dispatch(addProductToCart(cartProduct));
  };
};

export { addProductToCart, startAddProductToCart };
