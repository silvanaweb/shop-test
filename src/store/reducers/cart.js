const cartReducerDefaultState = [];

// cart product {
// image,
//   title,
//   price,
//   qty
// }

export default (state = cartReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_PRODUCT_TO_CART":
      console.log("current cart", state);
      return [...state, action.cartProduct];
    default:
      return state;
  }
};
