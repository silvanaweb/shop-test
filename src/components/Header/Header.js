import React from "react";
import { NavLink } from "react-router-dom";
import "./style.css";

const categories = ["trousers", "shorts", "polos", "gifts"];
const Header = () => {
  return (
    <header className="Header">
      <div className="content-container">
        <div className="Header__content">
          {categories.map(cat => (
            <NavLink key={cat} to={`/products/${cat}`} className="Header__link">
              {cat}
            </NavLink>
          ))}
        </div>
      </div>
    </header>
  );
};

export { Header };
