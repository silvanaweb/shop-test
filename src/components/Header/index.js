import React from "react";
import { connect } from "react-redux";
import { NavLink } from "react-router-dom";
import CartIcon from "../../assets/cart";
import "./style.css";

const categories = ["trousers", "jeans", "shorts", "polos", "gifts"];

const Header = ({ cart, history }) => {
  const cartColor = !!cart.length ? "#008080" : "#afd1d1";
  const cartDisabled = !!cart.length ? false : true;

  const onGoToCart = () => {
    history.push("/cart");
  };

  return (
    <header className="Header">
      <div className="content-container">
        <div className="Header__content">
          <div className="Header__left">
            {categories.map(cat => (
              <NavLink
                key={cat}
                to={`/products/${cat}`}
                className="Header__link"
              >
                {cat}
              </NavLink>
            ))}
          </div>
          <div className="Header__right">
            <button
              disabled={cartDisabled}
              className="Button--neutral "
              onClick={onGoToCart}
            >
              <CartIcon style={{ height: "4rem" }} color={cartColor} />
            </button>
          </div>
        </div>
      </div>
    </header>
  );
};

const mapStateToProps = (state, props) => {
  return {
    cart: state.cart
  };
};

export { Header };

export default connect(mapStateToProps)(Header);
