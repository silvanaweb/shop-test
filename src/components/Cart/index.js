import React from "react";
import { connect } from "react-redux";
import { startRemoveProductFromCart } from "../../store/actions/cart";
import { CartItem } from "./CartItem";
import "./style.css";

const calcTotal = cart =>
  cart.map(x => x.price * x.qty).reduce((acc, cv) => acc + cv, 0);

const Cart = ({ cart, startRemoveProductFromCart }) => {
  const onDeleteItemFromCart = (title, qty) => {
    startRemoveProductFromCart(title, qty);
  };

  return (
    <div>
      <h2>Cart</h2>
      {!!cart.length ? (
        <>
          <div>
            {cart.map(item => (
              <CartItem
                key={item.title}
                {...item}
                onDelete={onDeleteItemFromCart}
              />
            ))}
          </div>
          <div className="Cart__total">
            TOTAL <span>{calcTotal(cart)}</span>
            <button
              className="Button"
              style={{
                marginLeft: "2rem",
                paddingLeft: "4rem",
                paddingRight: "4rem"
              }}
            >
              Pay
            </button>
          </div>
        </>
      ) : (
        <h3>Your shopping cart is empty</h3>
      )}
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    cart: state.cart
  };
};

const mapDispatchToProps = dispatch => ({
  startRemoveProductFromCart: (title, qty) =>
    dispatch(startRemoveProductFromCart(title, qty))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Cart);
