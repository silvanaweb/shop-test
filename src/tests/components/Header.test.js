import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import React from "react";
import { categories, Header } from "../../components/Header";
import { cart } from "../fixtures/cart";

let history, wrapper;

beforeEach(() => {
  const cartData = [];
  history = { push: jest.fn() };
  wrapper = shallow(<Header cart={cartData} history={history} />);
});

test("should render Header with no cart correctly", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});
test("should render Header with cart correctly", () => {
  wrapper = shallow(<Header cart={cart} history={history} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should load correct links", () => {
  expect(wrapper.find(`.Header__link`).length).toBe(categories.length + 1);
  expect(
    wrapper
      .find("NavLink")
      .first()
      .props().to
  ).toBe("/products");
  const linkTos = ["/products", ...categories.map(x => `/products/${x}`)];
  wrapper.find(`.Header__link`).map((link, idx) => {
    expect(link.prop("to")).toBe(linkTos[idx]);
  });
});

test("should go to Cart page when Cart button is clicked", () => {
  wrapper
    .find('[data-test="CartButton"]')
    .first()
    .simulate("click");

  expect(history.push).toHaveBeenLastCalledWith("/cart");
});
