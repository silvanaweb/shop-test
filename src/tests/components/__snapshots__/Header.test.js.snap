// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header with cart correctly 1`] = `
<header
  className="Header"
>
  <div
    className="content-container"
  >
    <div
      className="Header__content"
    >
      <div
        className="Header__left"
      >
        <NavLink
          className="Header__link"
          exact={true}
          strict={true}
          to="/products"
        >
          All
        </NavLink>
        <NavLink
          className="Header__link"
          key="trousers"
          to="/products/trousers"
        >
          trousers
        </NavLink>
        <NavLink
          className="Header__link"
          key="jeans"
          to="/products/jeans"
        >
          jeans
        </NavLink>
        <NavLink
          className="Header__link"
          key="shorts"
          to="/products/shorts"
        >
          shorts
        </NavLink>
        <NavLink
          className="Header__link"
          key="polos"
          to="/products/polos"
        >
          polos
        </NavLink>
        <NavLink
          className="Header__link"
          key="gifts"
          to="/products/gifts"
        >
          gifts
        </NavLink>
      </div>
      <div
        className="Header__right"
      >
        <button
          className="Button--neutral"
          data-test="CartButton"
          onClick={[Function]}
        >
          <CartIcon
            color="#008080"
            style={
              Object {
                "height": "4rem",
              }
            }
          />
        </button>
      </div>
    </div>
  </div>
</header>
`;

exports[`should render Header with no cart correctly 1`] = `
<header
  className="Header"
>
  <div
    className="content-container"
  >
    <div
      className="Header__content"
    >
      <div
        className="Header__left"
      >
        <NavLink
          className="Header__link"
          exact={true}
          strict={true}
          to="/products"
        >
          All
        </NavLink>
        <NavLink
          className="Header__link"
          key="trousers"
          to="/products/trousers"
        >
          trousers
        </NavLink>
        <NavLink
          className="Header__link"
          key="jeans"
          to="/products/jeans"
        >
          jeans
        </NavLink>
        <NavLink
          className="Header__link"
          key="shorts"
          to="/products/shorts"
        >
          shorts
        </NavLink>
        <NavLink
          className="Header__link"
          key="polos"
          to="/products/polos"
        >
          polos
        </NavLink>
        <NavLink
          className="Header__link"
          key="gifts"
          to="/products/gifts"
        >
          gifts
        </NavLink>
      </div>
      <div
        className="Header__right"
      >
        <button
          className="Button--neutral"
          data-test="CartButton"
          onClick={[Function]}
        >
          <CartIcon
            color="#afd1d1"
            style={
              Object {
                "height": "4rem",
              }
            }
          />
        </button>
      </div>
    </div>
  </div>
</header>
`;
